aws_crt_common_ws/aws-crt-common/src/lib.rs

```diff
@@ -0,0 +1,596 @@
+    pub fn new_from_array(
+        allocator: &dyn NativeResource<CAllocator>,
+        str_data: &[u8],
+    ) -> Result<AwsString, i32> {
+        // assumes allocator is valid and str_data is not null.
+        // since the underlying allocation can fail, returns an error
+        // if aws_string_new_from_array() fails.
+        // The memory from this function is cleaned up in the destructor. If
+        // the destructor isn't run, the worse that will happen is a memory leak.
+        unsafe {
+            let c_str_ptr = aws_string_new_from_array(
+                allocator.get_native_resource(),
+                str_data.as_ptr(),
+                str_data.len(),
+            );
+            if c_str_ptr.is_null() {
+                return Err(last_error());
+            }
+
+            return Ok(AwsString {
+                c_str: c_str_ptr,
+                owned: true,
+            });
+        }
+    }
+
+    /// Returns the length of the string in bytes, not including any null terminators.
+    pub fn length(&self) -> usize {
+        // This is sound if you have an instance of this class.
+        unsafe {
+            return (*self.c_str).len;
+        }
+    }
+
+    /// Returns a pointer to an array of utf8 characters for the underlying String
+    pub fn bytes(&self) -> *const u8 {
+        // This is sound if you have an instance of this class.
+        unsafe {
+            return aws_string_bytes(self.c_str);
+        }
+    }
+
+    /// Returns a pointer to an array of utf8 characters as a c-string, including the null terminator.
+    /// **WARNING**
+    ///
+    /// It is the callers responsibility to ensure that the underlying memory is not freed too early.
+    ///
+    /// ```no_run
+    /// # #![allow(unused_must_use)]
+    /// // THIS IS UB & a use-after-free vulnerability!
+    /// use aws_crt_common::AwsString;
+    /// let allocator = todo!();
+    /// let ptr = AwsString::new_from_str(allocator, "Hello")
+    ///                       .expect("Creation of c string failed!")
+    ///                       .c_str();
+    /// unsafe {
+    ///     *ptr;
+    /// }
+    /// ```
+    ///
+    /// Instead, you must ensure that the AwsString lives long enough:
+    /// ```no_run
+    /// # #![allow(unused_must_use)]
+    /// use aws_crt_common::AwsString;
+    /// let allocator = todo!();
+    /// let aws_str = AwsString::new_from_str(allocator, "Hello")
+    ///                       .expect("Creation of c string failed!");
+    /// let ptr = aws_str.c_str();
+    /// unsafe {
+    ///     *ptr;
+    /// }
+    /// ```
+    pub fn c_str(&self) -> *const c_char {
+        // This is sound if you have an instance of this class.
+        unsafe {
+            return aws_string_bytes(self.c_str) as *const c_char;
+        }
+    }
+}
+
+impl ToString for AwsString {
+    fn to_string(&self) -> String {
+        // this is sound if you have an instance of this class and the original AwsString is valid utf-8
```

Comment:
is there somewhere that we guarantee that the `AwsString` can only be created from valid UTF-8? Otherwise this is UB.

Reply:
https://github.com/awslabs/aws-crt-rs/compare/russell/aws-string-safety

(remove non-owned variants of AwsString, they don't actually end up being used)